src/tir/transforms/local_pad.cc

```diff
@@ -0,0 +1,302 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <tvm/meta_schedule/postproc.h>
+#include <tvm/tir/op.h>
+#include <tvm/tir/stmt.h>
+#include <tvm/tir/stmt_functor.h>
+#include <tvm/tir/transform.h>
+
+#include <array>
+#include <utility>
+#include <vector>
+
+namespace tvm {
+namespace tir {
+namespace transform {
+namespace {
+
+/*!
+ * \brief Analyze the read and write accesses of the body statements, used by `LocalPadder`.
+ */
+class StorageAccessAnalyzer : public StmtExprVisitor {
+ private:
+  struct StorageType {
+    enum { kGlobal = 0, kShared, kLocal, kOthers };
+  };
+
+  void VisitStmt_(const BufferStoreNode* op) final {
+    write_marker_.SetStorageAccessMarker_(op->buffer);
+    StmtExprVisitor::VisitStmt_(op);
+  }
+  void VisitExpr_(const BufferLoadNode* op) final {
+    read_marker_.SetStorageAccessMarker_(op->buffer);
+    StmtExprVisitor::VisitExpr_(op);
+  }
+  class AccessMarker {
+   public:
+    void SetStorageAccessMarker_(const Buffer& buf) {
+      if (buf.scope() == "global") {
+        bit_vector_[StorageType::kGlobal] = true;
+      } else if (buf.scope() == "shared") {
+        bit_vector_[StorageType::kShared] = true;
+      } else if (buf.scope() == "local") {
+        bit_vector_[StorageType::kLocal] = true;
+      } else {
+        bit_vector_[StorageType::kOthers] = true;
+      }
+    }
+    bool NoAccesses() const {
+      return !(bit_vector_[StorageType::kGlobal] || bit_vector_[StorageType::kShared] ||
+               bit_vector_[StorageType::kLocal] || bit_vector_[StorageType::kOthers]);
+    }
+    bool OnlyGlobalAccesses() const {
+      return !(bit_vector_[StorageType::kShared] || bit_vector_[StorageType::kLocal] ||
+               bit_vector_[StorageType::kOthers]) &&
+             bit_vector_[StorageType::kGlobal];
+    }
+    bool OnlyLocalOrSharedAccesses() const {
+      return !(bit_vector_[StorageType::kGlobal] || bit_vector_[StorageType::kOthers]) &&
+             (bit_vector_[StorageType::kShared] || bit_vector_[StorageType::kLocal]);
+    }
+
+   private:
+    std::array<bool, StorageType::kOthers + 1> bit_vector_ = {false};
+  };
+  AccessMarker read_marker_, write_marker_;
+  std::pair<AccessMarker, AccessMarker> Analyze_(const Stmt& stmt) {
```

Comment:
ditto. Usually we put underscore after variables but not methods

Reply:
Remove underscore of private methods.